keps/sig-node/2837-pod-level-resource-spec/README.md

```diff
@@ -0,0 +1,1525 @@
+needs. Pod Overhead is considered to be added independently from resources
+specified in the Pod Spec by the kubelet.
+
+With the addition of pod-level requests and limits, there are a few ways to
+determine a pod's final resource allocation:
+
+* **[Preferred] Option 1: Pod Level Requests & Limits as an Override**
+
+  With this option, the resource requests and limits the users sets at the pod
+  level will directly determine the resource constraints applied to the pod's
+  cgroup. This means the pod's overall resource usage, including all its
+  contaienrs, will be capped by the limits you define at the pod level.
+  
+  For example, 
+  ```yaml
+  apiVersion: v1
+  kind: Pod
+  spec:
+    resources:
+      requests:
+        memory: 100Mi
+        cpu: 1
+      limits:
+        memory: 200Mi
+        cpu: 2
+    initContainers:
+      - name: init-container
+        image: busybox
+        ...    
+    containers:
+      - name: container-1
+        ...
+      - name: container-2
+        resources:
+          requests:
+            memory: 50Mi
+            cpu: 1
+  ```
+
+  **Resources requests at pod-level**
+
+  * The Scheduler will use the requested resources (CPU and memory) settings
+    specified at the pod level to find a node with enough available resources. 
+  * The Kubelet will use the CPU request at the pod-level resources to set 
+    cpu.weight in pod-level cgroup.
+
+  **Resources limits at pod-level**
+
+  * All init(Regular and Sidecar) and non-init containers will be capped by the
+    pod level limit which represents the combined peak usage of all containers.
+  * Kubelet will use CPU and memory limits from PodSpec to set cpu.max and
+    memory.max cgroup settings at pod-level respectively. 
+
+  **Pros**
+
+  * This approach is straightforward to understand and implement.
+  * It ensures enough resources are available for all containers.
+    within the pod, collectively
+  
+  **Cons**
+
+  * This method can sometimes lead to potential over-provisioning (resource wastage),
+    particularly when the init containers have significantly higher peak resource
+    usage compared to the non-init containers. In this case, allocating resources
+    based on the init container's needs can result in unnecessary resource
+    allocation for the application containers. However this can be mitigated by
+    setting container level resource spec for application containers.
+
+    For example, if your init container needs 2 CPU cores and 512Mi of memory
+    for a short burst during initialization, while your application container
+    only requires 0.5 CPU cores and 256Mi of memory. You should set pod level
+    requests and limits to 2 CPU and 512Mi of memory. Crucially, you could also
+    set container level requests and limits for your app container (0.5 CPU and
+    256Mi). This prevents the application container from using the extra
+    resources allocated to the pod, making the resources available for other workloads.
```

Comment:
This raises an interesting question for resize. A nice solution to this over-provisioning problem is to just resize the pod after it's initialized. But what if I want to resize the pod requests to be lower than the init container requests? Should we say that the init container is completed, so we disregard those requests? Or do we require the init container to be resized too? What if the node restarts, and the init containers are re-run? Food for thought...

Reply:
our guarantee today is that we can reinit containers at any moment. If we want to change it, we need to discuss